[Test Visibility] Go library (#26225)
DataDog/documentation: content/en/tests/test_impact_analysis/setup/go.md

---
title: Test Impact Analysis for Go
code_lang: go
type: multi-code-lang
code_lang_weight: 0
aliases:
  - /continuous_integration/intelligent_test_runner/go/
  - /continuous_integration/intelligent_test_runner/setup/go/
  - /intelligent_test_runner/setup/go
further_reading:
    - link: "/continuous_integration/tests"
      tag: "Documentation"
      text: "Explore Test Results and Performance"
    - link: "/continuous_integration/troubleshooting/"
      tag: "Documentation"
      text: "Troubleshooting CI Visibility"
---

## Compatibility

Test Impact Analysis is only supported on `orchestrion >= 0.9.5 (dd-trace-go >= 1.70.0)`.

## Setup

### Test Optimization

Prior to setting up Test Impact Analysis, set up [Test Optimization for Go][1]. If you are reporting data through the Agent, use v6.40 and later or v7.40 and later.

{{% ci-itr-activation-instructions %}}

## Run tests with Test Impact Analysis enabled

After completing setup, run your tests by using `go test` with the following code coverage options:

```bash
orchestrion go test ./... -cover -covermode=count -coverpkg ./...
```

> 1. `-cover`: The Test Impact Analysis feature uses the built-in Go's code coverage processor, so we need to enable code coverage collection in the `go test` command.

Reviewer: ```suggestion
1. `-cover`: The Test Impact Analysis feature uses the built-in Go's code coverage processor, so you need to enable code coverage collection in the `go test` command.
```

> 2. `-covermode`: must be either `count` or `atomic`. Because `set` is not supported, setting this value will disable test impact analysis.

Reviewer: ```suggestion
2. `-covermode`: must be either `count` or `atomic`. Because `set` is not supported, setting this value disables test impact analysis.
```

> 3. `-coverpkg`: the code coverage analysis for each test must be configured to apply in all package dependencies and not only for the package being tested. This way if a dependency changes we will be able to track the test affected by this change. If we are running the test command from the root of the project (where the go.mod file is) you can use the `./...` wildcard. If not, we have to manually list all package dependencies comma separated (`pattern1, pattern2, pattern3, ...`). For that, you could use the `go list ./...` command to get all the package names.

Reviewer: ```suggestion
3. `-coverpkg`: the code coverage analysis for each test must be configured to apply in all package dependencies and not only for the package being tested. This way, if a dependency changes, you can track the test affected by this change. If you run the test command from the root of the project (where the go.mod file is), you can use the `./...` wildcard. If not, you must manually list all package dependencies comma separated (`pattern1, pattern2, pattern3, ...`). For that, you could use the `go list ./...` command to get all the package names.
```

> <div class="alert alert-warning">Having an incorrect -coverpkg value will affect the ability of Test Impact Analysis to correctly track test coverage.</div>

Reviewer: ```suggestion
<div class="alert alert-warning">Having an incorrect -coverpkg value affects the ability of Test Impact Analysis to correctly track test coverage.</div>
```
